### ISSUE TITLE
[RFC] special metavariable syntax for dataflow

### ISSUE THREAD

**Issue by nbrahms** [semgrep/semgrep#3085]
## Motivation

We are beginning to add dataflow support; viz., we have recently added constant propagation support.

For example, we might use the pattern:

    foo(1)

to match the code:

    x = 1;
    foo(x);

However, we are often interested not in the _value_ of the constant itself, but whether an object is _effectively_ constant at its call site; or otherwise trustworthy.

To that end, I suggest adding "special" metavariables. Each special metavariable would be an uppercase token preceded by `$$`.

## Proposal

I envision starting with the following:

- `$$CONSTANT`: A value known to be a constant literal
- `$$NON_CONSTANT`: A value not known to be a constant
- `$$LOCAL_EXPRESSION`: A value known to be an expression that depends on local scope only
- `$$NON_LOCAL_EXPRESSION`: A value not known to depend on local scope only

## Some examples

All matches are to `foo($$SPECIAL_VARIABLE)`.

`$$CONSTANT` matches:

    x = 1
    foo(x)

    x = "a"
    foo(x)

    foo("a" + 3)

    foo(f"hello {'world'})

    for (x in [0, 1, 2, 3]) {
         foo(x)
    }

`$$NON_CONSTANT` matches:

    x = readline("enter a number")
    foo(x)

    def bar(x) {
        foo(x)
    }

    def bar() {
        x = do_something([1, 2, 3, 4]);
        foo(x)
    }

    def bar(index) {
        whitelisted_values = ["a", "b", "c"];
        foo(whitelisted_values[index]);
    }

`$$LOCAL_EXPRESSION`:

    def bar() {
        x = do_something([1, 2, 3, 4]);
        foo(x)
    }

`$$NON_LOCAL_EXPRESSION`:

    def bar(x) {
        foo(x)
    }

    def bar(index) {
        whitelisted_values = ["a", "b", "c"];
        foo(whitelisted_values[index]);
    }

## Use cases

The most straightforward use case is to detect injection. All of the following patterns reduce a large swath of expensive injection checks:

    eval($$NON_CONSTANT)

    query.execute($$NON_CONSTANT)

    template.render($$NON_CONSTANT)

Note we can still get false positives (e.g. in the "whitelisted_values" example above).

**Comment by minusworld** [comment]
This looks awesome. I think my only feedback is that `$$NON_LOCAL_EXPRESSION` does not necessarily make clear what it is. Something like `$$ANY_SCOPE_EXPRESSION` might be better

**Comment by mschwager** [comment]
> The most straightforward use case is to detect injection. All of the following patterns reduce a large swath of expensive injection checks ...

If the primary use-case is injection, would we be better served by improving our taint analysis? Improving taint seems like a more generalized solution, and doesn't further complicate the pattern syntax.

**Comment by ievans** [comment]
There's some prior discussion in https://github.com/returntocorp/semgrep/issues/1645 where I had proposed that rather than introducing new operators we could use the existing language's type system ability to specify const (if it exists), eg `foo($X: const<int>)`. I think that proposal might be more intuitive in terms of syntax, but is worse for advanced users as most won't care about the specific type that is const.

👍 to @mschwager suggestion about improving taint mode being preferable to new syntax.

**Comment by IagoAbal** [comment]
Nice proposal!

Perhaps it could be `$$CONSTANT_[A-Z_0-9]+` (e.g. `$$CONSTANT_X`) instead so that you can treat it as yet another metavariable, simply one that only matches conditionally.

I also like @ievans proposal, and one could accept type wildcards `_` as in `const<_>` (or maybe `const<...>` to reuse the very familiar `...`?). Although this doesn't cover the `$$LOCAL_EXPRESSION` proposal.

**Comment by IagoAbal** [comment]
Also :+1: to improving taint analysis although I think this is also a nice-have. For taint analysis we also need a concrete proposal like this one.

**Comment by aryx** [comment]
regarding constant, this could be done by typed metavariables? $CST : int, or $CST : int_constant?

**Comment by aryx** [comment]
We could put more things in typed metavariables, the type could be a property of the expression (being constant, local, etc.)
Actually in coccinelle we were declaring metavariables separately of the pattern, as in
@@
expr $X;
int $Y;
@@
foo($X, $Y);

**Comment by aryx** [comment]
Just like we have metavariable-regexp to restraint a metavariable, we could have another construct to restraint the type/constness/tainted/... property of a metavariable.

**Comment by emjin** [comment]
As a related question, I had initially only imagined adding typed metavariables for other languages once we had typing information, but the eslint rule experience made me realize that it might be useful more broadly. Languages like Python would only be able to understand the types of constants, but that might be enough to justify adding the feature. @minusworld and other security devs for whether they would like this.

**Comment by clintgibler** [comment]
This is super exciting, thanks for creating this @nbrahms! 🙏 

+1 @mschwager that improving our taint analysis would be great, though I think there's value in `$$NON_CONSTANT` regardless, as oftentimes that's what rule authors care about, even if they don't have the time/energy to flesh out all the relevant sinks/sources of what they're looking for. 

I think a relatively common use case (from me as well as a few users) is "OK this function is dangerous, show me everywhere`this arg` is not a constant" - this iterative, quick to start workflow seems more difficult if the only way you could do it is with taint mode.

I agree with the concern about adding more operators and making Semgrep more complex to use :/ 

👍  @minusworld  that `$$NON_LOCAL_EXPRESSION` isn't obvious what it means, though I can't think of a better term right now.

<hr>

**tl;dr**: This sounds great, `$$NON_CONSTANT` and `$$NON_LOCAL_EXPRESSION` seem super useful (even with taint mode), can we think of a clearer name for `$$NON_LOCAL_EXPRESSION`, and how can we keep Semgrep simple?

**Comment by nbrahms** [comment]
Perhaps an "alternate" proposal, based on the feedback here, would be to borrow type modifiers from TypeScript, and then have something like the following (copying from the original RFC):

## Proposal

We would update type annotations so that a type could be formed with type modifiers, literal types, and type wildcards:

    [modifier ...] (type | '_')

I envision adding the following type modifiers:

- `static` (or `const`?): The metavariable is known to be a static constant
- `dynamic` (or `nonconst`?): The metavariable is not known to be a static constant

- `local`: The metavariable is known to be independent of data external to the scope
- `extern`: The metavariable is not known to be independent of data external to the scope

Note that `static` and `dynamic` should be exclusive, as should `local` and `extern`. Alternatively, we could keep just `static` and `local`, and have a meta-modifier (say `!`), so that `dynamic` becomes `!static`. In my opinion this is a bit clumsy, however.

And type wildcard:
- `_`: Indicates an unknown type

## Some examples

`foo($X: static _)` matches:

    x = 1
    foo(x)

    x = "a"
    foo(x)

    foo("a" + 3)

    foo(f"hello {'world'})

    for (x in [0, 1, 2, 3]) {
         foo(x)
    }

`foo($X: dynamic _)` matches:

    x = readline("enter a number")
    foo(x)

    def bar(x) {
        foo(x)
    }

    def bar() {
        x = do_something([1, 2, 3, 4]);
        foo(x)
    }

    def bar(index) {
        whitelisted_values = ["a", "b", "c"];
        foo(whitelisted_values[index]);
    }

`foo($X: local _)`:

    def bar() {
        x = do_something([1, 2, 3, 4]);
        foo(x)
    }

`foo($X: extern _)`:

    def bar(x) {
        foo(x)
    }

    def bar(index) {
        whitelisted_values = ["a", "b", "c"];
        foo(whitelisted_values[index]);
    }

## Use cases

The most straightforward use case is to detect injection. All of the following patterns reduce a large swath of expensive injection checks:

    eval($X: extern _)

    query.execute($X: extern _)

    template.render($X: extern _)

Note we can still get false positives (e.g. in the "whitelisted_values" example above).

## Equivalent searches

We could also remove the need for `dynamic` and `extern` by using pattern composition:

(in the v1 API)

    patterns:
      - pattern: foo($X)
      - pattern-not: foo($X: local _)

(in the proposed boolean API)

    match:
      - pattern: foo($X)
      - not:
          pattern: foo($X: local _)

**Comment by IagoAbal** [comment]
Maybe we should move this to Quip? Looks like a decent Hack Week project.

**Comment by IagoAbal** [comment]
I think https://github.com/returntocorp/semgrep/issues/3085#issuecomment-844237454 is good, we just need to make a decision whether this is still relevant and what is it's priority.
